diff --git a/phase3/blinky.c b/phase3/blinky.c
index 26282d6..a44adf3 100644
--- a/phase3/blinky.c
+++ b/phase3/blinky.c
@@ -16,5 +16,12 @@ static void spin_sleep_ms(unsigned int ms) {
 
 int main(void) {
   // STEP 1: Set GPIO Pin 16 as output.
+  *GPIO_FSEL1 = 0b001 << 18;
   // STEP 2: Continuously set and clear GPIO 16.
+  while (1) {
+    *GPIO_SET0 = 1 << 16;
+    spin_sleep_ms(1000);
+    *GPIO_CLR0 = 1 << 16;
+    spin_sleep_ms(1000);
+  }
 }
diff --git a/phase4/.vscode/settings.json b/phase4/.vscode/settings.json
index e69de29..2dd73f2 100644
--- a/phase4/.vscode/settings.json
+++ b/phase4/.vscode/settings.json
@@ -0,0 +1,4 @@
+{
+    "rust.target": "aarch64-none-elf",
+    "rust.sysroot": "/Users/macbookair/.xargo"
+}
\ No newline at end of file
diff --git a/phase4/src/lib.rs b/phase4/src/lib.rs
index 8019dbf..f3c9b5d 100644
--- a/phase4/src/lib.rs
+++ b/phase4/src/lib.rs
@@ -22,5 +22,12 @@ fn spin_sleep_ms(ms: usize) {
 #[no_mangle]
 pub unsafe extern "C" fn kmain() {
     // STEP 1: Set GPIO Pin 16 as output.
+    GPIO_FSEL1.write_volatile(1 << 18);
     // STEP 2: Continuously set and clear GPIO 16.
+    loop {
+        GPIO_SET0.write_volatile(1 << 16);
+        spin_sleep_ms(1000);
+        GPIO_CLR0.write_volatile(1 << 16);
+        spin_sleep_ms(1000);
+    }
 }
