diff --git a/questions/arm-el b/questions/arm-el
index e69de29..3114475 100644
--- a/questions/arm-el
+++ b/questions/arm-el
@@ -0,0 +1,3 @@
+mrs x0, CurrentEL
+and x0, x0, #0b1100
+lsr x0, x0, #2
\ No newline at end of file
diff --git a/questions/arm-pc b/questions/arm-pc
index e69de29..1d2446e 100644
--- a/questions/arm-pc
+++ b/questions/arm-pc
@@ -0,0 +1,6 @@
+adr/ldr lr, A
+ret
+
+adr/ldr x0, A
+msr ELR_ELx, x0
+eret
diff --git a/questions/arm-sp-el b/questions/arm-sp-el
index e69de29..0ee42a3 100644
--- a/questions/arm-sp-el
+++ b/questions/arm-sp-el
@@ -0,0 +1,4 @@
+adr/ldr x0, B
+msr SP_ELn, x0
+mov x0, 1
+msr SPSR_ELx, x0
\ No newline at end of file
diff --git a/questions/arm-x30 b/questions/arm-x30
index e69de29..8b40903 100644
--- a/questions/arm-x30
+++ b/questions/arm-x30
@@ -0,0 +1 @@
+lr and w30.
