diff --git a/bootloader/.vscode/settings.json b/bootloader/.vscode/settings.json
index e69de29..2dd73f2 100644
--- a/bootloader/.vscode/settings.json
+++ b/bootloader/.vscode/settings.json
@@ -0,0 +1,4 @@
+{
+    "rust.target": "aarch64-none-elf",
+    "rust.sysroot": "/Users/macbookair/.xargo"
+}
\ No newline at end of file
diff --git a/bootloader/Xargo.toml b/bootloader/Xargo.toml
index 5e8c094..3fb0b62 100644
--- a/bootloader/Xargo.toml
+++ b/bootloader/Xargo.toml
@@ -1,6 +1,7 @@
 [dependencies]
 core = {}
 std_unicode = {}
+alloc = {}
 
 [dependencies.compiler_builtins]
 features = ["mem"]
diff --git a/bootloader/src/allocator b/bootloader/src/allocator
index e69de29..eb56ef6 120000
--- a/bootloader/src/allocator
+++ b/bootloader/src/allocator
@@ -0,0 +1 @@
+../../kernel/src/allocator
\ No newline at end of file
diff --git a/bootloader/src/console.rs b/bootloader/src/console.rs
index e69de29..9bf5d6e 120000
--- a/bootloader/src/console.rs
+++ b/bootloader/src/console.rs
@@ -0,0 +1 @@
+../../kernel/src/console.rs
\ No newline at end of file
diff --git a/bootloader/src/kmain.rs b/bootloader/src/kmain.rs
index 591a219..ba6170e 100644
--- a/bootloader/src/kmain.rs
+++ b/bootloader/src/kmain.rs
@@ -1,10 +1,30 @@
 #![feature(asm, lang_items)]
+#![feature(alloc, allocator_api, global_allocator)]
+#![feature(optin_builtin_traits)]
+#![feature(const_fn)]
+#![feature(decl_macro)]
+#![feature(core_intrinsics)]
 
-extern crate xmodem;
+extern crate alloc;
 extern crate pi;
+extern crate xmodem;
 
+use std::io;
+use std::io::Write;
+use pi::uart::MiniUart;
+
+pub mod mutex;
+pub mod console;
+pub mod allocator;
 pub mod lang_items;
 
+#[cfg(not(test))]
+use allocator::Allocator;
+
+#[cfg(not(test))]
+#[global_allocator]
+pub static ALLOCATOR: Allocator = Allocator::uninitialized();
+
 /// Start address of the binary to load and of the bootloader.
 const BINARY_START_ADDR: usize = 0x80000;
 const BOOTLOADER_START_ADDR: usize = 0x4000000;
@@ -19,11 +39,31 @@ const MAX_BINARY_SIZE: usize = BOOTLOADER_START_ADDR - BINARY_START_ADDR;
 fn jump_to(addr: *mut u8) -> ! {
     unsafe {
         asm!("br $0" : : "r"(addr as usize));
-        loop { asm!("nop" :::: "volatile")  }
+        loop {
+            asm!("nop" :::: "volatile")
+        }
     }
 }
 
 #[no_mangle]
 pub extern "C" fn kmain() {
     // FIXME: Implement the bootloader.
+    let mut uart = MiniUart::new();
+    uart.set_read_timeout(750);
+
+    loop {
+        let dest = unsafe { std::slice::from_raw_parts_mut(BINARY_START, MAX_BINARY_SIZE) };
+        match xmodem::Xmodem::receive(&mut uart, io::Cursor::new(dest)) {
+            Ok(_) => {
+                // Succeed
+                jump_to(BINARY_START)
+            }
+            Err(err) => match err.kind() {
+                io::ErrorKind::TimedOut => continue,
+                io::ErrorKind::InvalidData => continue, // might receive 0x00 when no input
+                _ => uart.write_fmt(format_args!("Error: {:?}\r\n", err))
+                    .unwrap(),
+            },
+        }
+    }
 }
diff --git a/bootloader/src/mutex.rs b/bootloader/src/mutex.rs
index e69de29..a29af40 120000
--- a/bootloader/src/mutex.rs
+++ b/bootloader/src/mutex.rs
@@ -0,0 +1 @@
+../../kernel/src/mutex.rs
\ No newline at end of file
diff --git a/kernel/.vscode/settings.json b/kernel/.vscode/settings.json
index e69de29..2dd73f2 100644
--- a/kernel/.vscode/settings.json
+++ b/kernel/.vscode/settings.json
@@ -0,0 +1,4 @@
+{
+    "rust.target": "aarch64-none-elf",
+    "rust.sysroot": "/Users/macbookair/.xargo"
+}
\ No newline at end of file
diff --git a/kernel/Cargo.toml b/kernel/Cargo.toml
index e2aee8b..dbe2f3c 100644
--- a/kernel/Cargo.toml
+++ b/kernel/Cargo.toml
@@ -9,7 +9,6 @@ path = "src/kmain.rs"
 
 [profile.dev]
 panic = "abort"
-lto = true
 debug = true
 
 [profile.test]
diff --git a/kernel/src/allocator/bin.rs b/kernel/src/allocator/bin.rs
index 28384f2..3daae5b 100644
--- a/kernel/src/allocator/bin.rs
+++ b/kernel/src/allocator/bin.rs
@@ -1,19 +1,40 @@
-use std::fmt;
+use std::cmp::max;
+use std::cmp::min;
 use alloc::heap::{AllocErr, Layout};
 
 use allocator::util::*;
 use allocator::linked_list::LinkedList;
+use std::mem::size_of;
+
+const SIZEOF_USIZE: usize = size_of::<usize>();
 
 /// A simple allocator that allocates based on size classes.
+#[derive(Debug)]
 pub struct Allocator {
     // FIXME: Add the necessary fields.
+    // minimum=usize=8B, 2^3B
+    free_list: [LinkedList; 32],
 }
 
 impl Allocator {
     /// Creates a new bin allocator that will allocate memory from the region
     /// starting at address `start` and ending at address `end`.
     pub fn new(start: usize, end: usize) -> Allocator {
-        unimplemented!("bin allocator")
+        let mut free_list = [LinkedList::new(); 32];
+        let mut current_start = start;
+
+        while current_start < end {
+            let lowbit = current_start & (!current_start + 1);
+            let size = max(min(lowbit, prev_power_of_two(end - current_start)), SIZEOF_USIZE);
+            unsafe {
+                free_list[size.trailing_zeros() as usize].push(current_start as *mut usize);
+            }
+            current_start += size;
+        }
+
+        Allocator {
+            free_list: free_list,
+        }
     }
 
     /// Allocates memory. Returns a pointer meeting the size and alignment
@@ -37,7 +58,31 @@ impl Allocator {
     /// (`AllocError::Exhausted`) or `layout` does not meet this allocator's
     /// size or alignment constraints (`AllocError::Unsupported`).
     pub fn alloc(&mut self, layout: Layout) -> Result<*mut u8, AllocErr> {
-        unimplemented!("bin allocation")
+        let size = max(
+            layout.size().next_power_of_two(),
+            max(layout.align(), SIZEOF_USIZE),
+        );
+        let class = size.trailing_zeros() as usize;
+        for i in class..self.free_list.len() {
+            if !self.free_list[i].is_empty() {
+                for j in (class + 1..i + 1).rev() {
+                    let block = self.free_list[j]
+                        .pop()
+                        .expect("bigger block should have free space");
+                    unsafe {
+                        self.free_list[j - 1].push((block as usize + (1 << (j - 1))) as *mut usize);
+                        self.free_list[j - 1].push(block);
+                    }
+                }
+
+                let result = Ok(self.free_list[class]
+                    .pop()
+                    .expect("current block should have free space now")
+                    as *mut u8);
+                return result;
+            }
+        }
+        Err(AllocErr::Exhausted { request: layout })
     }
 
     /// Deallocates the memory referenced by `ptr`.
@@ -54,6 +99,36 @@ impl Allocator {
     /// Parameters not meeting these conditions may result in undefined
     /// behavior.
     pub fn dealloc(&mut self, ptr: *mut u8, layout: Layout) {
-        unimplemented!("bin deallocation")
+        let size = max(
+            layout.size().next_power_of_two(),
+            max(layout.align(), SIZEOF_USIZE),
+        );
+        let class = size.trailing_zeros() as usize;
+
+        unsafe {
+            self.free_list[class].push(ptr as *mut usize);
+            let mut current_ptr = ptr as usize;
+            let mut current_class = class;
+            loop {
+                let buddy = current_ptr ^ (1 << current_class);
+                let mut flag = false;
+                for block in self.free_list[current_class].iter_mut() {
+                    if block.value() as usize == buddy {
+                        block.pop();
+                        flag = true;
+                        break;
+                    }
+                }
+                if flag {
+                    self.free_list[current_class].pop();
+                    current_ptr = min(current_ptr, buddy);
+                    current_class += 1;
+                    self.free_list[current_class]
+                        .push(current_ptr as *mut usize);
+                } else {
+                    break;
+                }
+            }
+        }
     }
 }
diff --git a/kernel/src/allocator/bump.rs b/kernel/src/allocator/bump.rs
index e8741cf..dd66933 100644
--- a/kernel/src/allocator/bump.rs
+++ b/kernel/src/allocator/bump.rs
@@ -13,7 +13,10 @@ impl Allocator {
     /// Creates a new bump allocator that will allocate memory from the region
     /// starting at address `start` and ending at address `end`.
     pub fn new(start: usize, end: usize) -> Allocator {
-        unimplemented!("bump allocator")
+        Allocator {
+            current: start,
+            end: end,
+        }
     }
 
     /// Allocates memory. Returns a pointer meeting the size and alignment
@@ -37,7 +40,23 @@ impl Allocator {
     /// (`AllocError::Exhausted`) or `layout` does not meet this allocator's
     /// size or alignment constraints (`AllocError::Unsupported`).
     pub fn alloc(&mut self, layout: Layout) -> Result<*mut u8, AllocErr> {
-        unimplemented!("bump allocation")
+        let new_start = align_up(self.current, layout.align()).saturating_add(layout.size());
+
+        if new_start >= self.end {
+            return Err(AllocErr::Exhausted { request: layout });
+        }
+
+        if cfg!(dev) {
+            use console::noblock_kprintln;
+            noblock_kprintln!(
+                "Allocating buffer from {} to {}",
+                new_start - layout.size(),
+                new_start
+            );
+        }
+
+        self.current = new_start;
+        Ok((new_start - layout.size()) as *mut u8)
     }
 
     /// Deallocates the memory referenced by `ptr`.
@@ -53,7 +72,15 @@ impl Allocator {
     ///
     /// Parameters not meeting these conditions may result in undefined
     /// behavior.
-    pub fn dealloc(&mut self, _ptr: *mut u8, _layout: Layout) {
-        unimplemented!("bump deallocation")
+    pub fn dealloc(&mut self, ptr: *mut u8, layout: Layout) {
+        // Do nothing
+        if cfg!(dev) {
+            use console::noblock_kprintln;
+            noblock_kprintln!(
+                "Deallocating buffer from {} to {}",
+                ptr as usize,
+                ptr as usize + layout.size()
+            );
+        }
     }
 }
diff --git a/kernel/src/allocator/mod.rs b/kernel/src/allocator/mod.rs
index f82206d..95c44b7 100644
--- a/kernel/src/allocator/mod.rs
+++ b/kernel/src/allocator/mod.rs
@@ -10,6 +10,7 @@ mod tests;
 use mutex::Mutex;
 use alloc::heap::{Alloc, AllocErr, Layout};
 use std::cmp::max;
+use pi::atags::Atags;
 
 /// Thread-safe (locking) wrapper around a particular memory allocator.
 #[derive(Debug)]
@@ -57,7 +58,11 @@ unsafe impl<'a> Alloc for &'a Allocator {
     /// (`AllocError::Exhausted`) or `layout` does not meet this allocator's
     /// size or alignment constraints (`AllocError::Unsupported`).
     unsafe fn alloc(&mut self, layout: Layout) -> Result<*mut u8, AllocErr> {
-        self.0.lock().as_mut().expect("allocator uninitialized").alloc(layout)
+        self.0
+            .lock()
+            .as_mut()
+            .expect("allocator uninitialized")
+            .alloc(layout)
     }
 
     /// Deallocates the memory referenced by `ptr`.
@@ -74,7 +79,11 @@ unsafe impl<'a> Alloc for &'a Allocator {
     /// Parameters not meeting these conditions may result in undefined
     /// behavior.
     unsafe fn dealloc(&mut self, ptr: *mut u8, layout: Layout) {
-        self.0.lock().as_mut().expect("allocator uninitialized").dealloc(ptr, layout);
+        self.0
+            .lock()
+            .as_mut()
+            .expect("allocator uninitialized")
+            .dealloc(ptr, layout);
     }
 }
 
@@ -87,7 +96,21 @@ extern "C" {
 ///
 /// This function is expected to return `Some` under all normal cirumstances.
 fn memory_map() -> Option<(usize, usize)> {
-    let binary_end = unsafe { (&_end as *const u8) as u32 };
+    let binary_end = unsafe { (&_end as *const u8) as usize };
 
-    unimplemented!("memory map fetch")
+    for tag in Atags::get() {
+        match tag.mem() {
+            Some(mem) => {
+                let mut start = mem.start as usize;
+                let end = (mem.start + mem.size) as usize;
+                if binary_end < end {
+                    start = max(start, binary_end);
+                }
+                return Some((start, end));
+            }
+            _ => {}
+        }
+    }
+
+    None
 }
diff --git a/kernel/src/allocator/util.rs b/kernel/src/allocator/util.rs
index 11e0477..c96a7c4 100644
--- a/kernel/src/allocator/util.rs
+++ b/kernel/src/allocator/util.rs
@@ -1,3 +1,5 @@
+use std::intrinsics::size_of;
+
 /// Align `addr` downwards to the nearest multiple of `align`.
 ///
 /// The returned usize is always <= `addr.`
@@ -6,7 +8,11 @@
 ///
 /// Panics if `align` is not a power of 2.
 pub fn align_down(addr: usize, align: usize) -> usize {
-    unimplemented!()
+    if !align.is_power_of_two() {
+        panic!("align_down: align not a power of 2");
+    }
+
+    addr & !(align - 1)
 }
 
 /// Align `addr` upwards to the nearest multiple of `align`.
@@ -17,5 +23,9 @@ pub fn align_down(addr: usize, align: usize) -> usize {
 ///
 /// Panics if `align` is not a power of 2.
 pub fn align_up(addr: usize, align: usize) -> usize {
-    unimplemented!()
+    align_down(addr.saturating_add(align - 1), align)
+}
+
+pub fn prev_power_of_two(num: usize) -> usize {
+    unsafe { 1 << (8 * (size_of::<usize>()) - num.leading_zeros() as usize - 1) }
 }
diff --git a/kernel/src/console.rs b/kernel/src/console.rs
index e0f18f9..35609c1 100644
--- a/kernel/src/console.rs
+++ b/kernel/src/console.rs
@@ -7,7 +7,7 @@ use mutex::Mutex;
 
 /// A global singleton allowing read/write access to the console.
 pub struct Console {
-    inner: Option<MiniUart>
+    inner: Option<MiniUart>,
 }
 
 impl Console {
@@ -19,35 +19,36 @@ impl Console {
     /// Initializes the console if it's not already initialized.
     #[inline]
     fn initialize(&mut self) {
-        unimplemented!()
+        self.inner.get_or_insert_with(&MiniUart::new);
     }
 
     /// Returns a mutable borrow to the inner `MiniUart`, initializing it as
     /// needed.
     fn inner(&mut self) -> &mut MiniUart {
-        unimplemented!()
+        self.initialize();
+        self.inner.as_mut().unwrap()
     }
 
     /// Reads a byte from the UART device, blocking until a byte is available.
     pub fn read_byte(&mut self) -> u8 {
-        unimplemented!()
+        self.inner().read_byte()
     }
 
     /// Writes the byte `byte` to the UART device.
     pub fn write_byte(&mut self, byte: u8) {
-        unimplemented!()
+        self.inner().write_byte(byte)
     }
 }
 
 impl io::Read for Console {
     fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
-        unimplemented!()
+        self.inner().read(buf)
     }
 }
 
 impl io::Write for Console {
     fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
-        unimplemented!()
+        self.inner().write(buf)
     }
 
     fn flush(&mut self) -> io::Result<()> {
@@ -57,7 +58,7 @@ impl io::Write for Console {
 
 impl fmt::Write for Console {
     fn write_str(&mut self, s: &str) -> fmt::Result {
-        unimplemented!()
+        self.inner().write_str(s)
     }
 }
 
@@ -89,3 +90,23 @@ pub macro kprintln {
 pub macro kprint($($arg:tt)*) {
     _print(format_args!($($arg)*))
 }
+
+/// Internal function called by the `noblock_kprint[ln]!` macros.
+#[doc(hidden)]
+pub fn _noblock_print(args: fmt::Arguments) {
+    use std::fmt::Write;
+    let mut console = Console::new();
+    console.write_fmt(args).unwrap();
+}
+
+/// Like `println!`, but non-blocking and for kernel-space.
+pub macro noblock_kprintln {
+    () => (noblock_kprint!("\n")),
+    ($fmt:expr) => (noblock_kprint!(concat!($fmt, "\n"))),
+    ($fmt:expr, $($arg:tt)*) => (noblock_kprint!(concat!($fmt, "\n"), $($arg)*))
+}
+
+/// Like `print!`, but non-blocking and for kernel-space.
+pub macro noblock_kprint($($arg:tt)*) {
+    _noblock_print(format_args!($($arg)*))
+}
diff --git a/kernel/src/kmain.rs b/kernel/src/kmain.rs
index 471db36..fedb2d7 100644
--- a/kernel/src/kmain.rs
+++ b/kernel/src/kmain.rs
@@ -4,7 +4,6 @@
 #![feature(asm)]
 #![feature(optin_builtin_traits)]
 #![feature(decl_macro)]
-#![feature(repr_align)]
 #![feature(attr_literals)]
 #![feature(exclusive_range_pattern)]
 #![feature(alloc, allocator_api, global_allocator)]
@@ -16,6 +15,12 @@ extern crate pi;
 extern crate stack_vec;
 
 pub mod allocator;
+use pi::gpio::Gpio;
+use std::fmt::Write;
+use pi::uart::MiniUart;
+use pi::timer::spin_sleep_ms;
+use pi::timer::current_time;
+
 pub mod lang_items;
 pub mod mutex;
 pub mod console;
@@ -28,8 +33,52 @@ use allocator::Allocator;
 #[global_allocator]
 pub static ALLOCATOR: Allocator = Allocator::uninitialized();
 
+use console::{kprintln, CONSOLE};
+
+fn blinky() {
+    let pin16 = Gpio::new(16);
+    let mut pin_out16 = pin16.into_output();
+
+    loop {
+        pin_out16.set();
+        spin_sleep_ms(1000);
+        pin_out16.clear();
+        spin_sleep_ms(1000);
+    }
+}
+
+fn echo() {
+    let mut uart = MiniUart::new();
+
+    loop {
+        let byte = uart.read_byte();
+        uart.write_byte(byte);
+    }
+}
+
+fn time() {
+    loop {
+        kprintln!("Current time: {}", current_time());
+        spin_sleep_ms(1000);
+    }
+}
+
+fn hex() {
+    let mut console = CONSOLE.lock();
+    loop {
+        let byte = console.read_byte();
+        console.write_fmt(format_args!("{:#X}", byte)).unwrap();
+    }
+}
+
 #[no_mangle]
 #[cfg(not(test))]
 pub extern "C" fn kmain() {
     ALLOCATOR.initialize();
+    // FIXME: Start the shell.
+    // blinky();
+    // echo();
+    // time();
+    // hex();
+    shell::shell("> ");
 }
diff --git a/kernel/src/lang_items.rs b/kernel/src/lang_items.rs
index 3a1e46f..83848e3 100644
--- a/kernel/src/lang_items.rs
+++ b/kernel/src/lang_items.rs
@@ -3,8 +3,15 @@
 #[lang = "panic_fmt"]
 pub extern fn panic_fmt(fmt: ::std::fmt::Arguments, file: &'static str, line: u32, col: u32) -> ! {
     // FIXME: Print `fmt`, `file`, and `line` to the console.
+    // Avoid deadlock
+    use console::noblock_kprintln;
+    noblock_kprintln!("Kernel Panic at file {}\n\t\tline {}, col {}: {}", file, line, col, fmt);
 
-    loop { unsafe { asm!("wfe") } }
+    loop {
+        unsafe { asm!("wfe") }
+    }
 }
 
-#[cfg(not(test))] #[lang = "eh_personality"] pub extern fn eh_personality() {}
+#[cfg(not(test))]
+#[lang = "eh_personality"]
+pub extern "C" fn eh_personality() {}
diff --git a/kernel/src/mutex.rs b/kernel/src/mutex.rs
index e52b172..394015f 100644
--- a/kernel/src/mutex.rs
+++ b/kernel/src/mutex.rs
@@ -1,6 +1,6 @@
 use std::sync::atomic::{AtomicBool, Ordering};
 use std::cell::UnsafeCell;
-use std::ops::{DerefMut, Deref, Drop};
+use std::ops::{Deref, DerefMut, Drop};
 use std::fmt;
 
 #[repr(align(32))]
@@ -9,21 +9,21 @@ pub struct Mutex<T> {
     lock: AtomicBool,
 }
 
-unsafe impl<T: Send> Send for Mutex<T> { }
-unsafe impl<T: Send> Sync for Mutex<T> { }
+unsafe impl<T: Send> Send for Mutex<T> {}
+unsafe impl<T: Send> Sync for Mutex<T> {}
 
 pub struct MutexGuard<'a, T: 'a> {
-    lock: &'a Mutex<T>
+    lock: &'a Mutex<T>,
 }
 
-impl<'a, T> !Send for MutexGuard<'a, T> { }
-unsafe impl<'a, T: Sync> Sync for MutexGuard<'a, T> { }
+impl<'a, T> !Send for MutexGuard<'a, T> {}
+unsafe impl<'a, T: Sync> Sync for MutexGuard<'a, T> {}
 
 impl<T> Mutex<T> {
     pub const fn new(val: T) -> Mutex<T> {
         Mutex {
             lock: AtomicBool::new(false),
-            data: UnsafeCell::new(val)
+            data: UnsafeCell::new(val),
         }
     }
 }
@@ -48,7 +48,7 @@ impl<T> Mutex<T> {
         loop {
             match self.try_lock() {
                 Some(guard) => return guard,
-                None => continue
+                None => continue,
             }
         }
     }
@@ -62,7 +62,7 @@ impl<'a, T: 'a> Deref for MutexGuard<'a, T> {
     type Target = T;
 
     fn deref(&self) -> &T {
-        unsafe { & *self.lock.data.get() }
+        unsafe { &*self.lock.data.get() }
     }
 }
 
@@ -82,7 +82,7 @@ impl<T: fmt::Debug> fmt::Debug for Mutex<T> {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
         match self.try_lock() {
             Some(guard) => f.debug_struct("Mutex").field("data", &&*guard).finish(),
-            None => f.debug_struct("Mutex").field("data", &"<locked>").finish()
+            None => f.debug_struct("Mutex").field("data", &"<locked>").finish(),
         }
     }
 }
diff --git a/kernel/src/shell.rs b/kernel/src/shell.rs
index 03fb8fb..5b66761 100644
--- a/kernel/src/shell.rs
+++ b/kernel/src/shell.rs
@@ -1,16 +1,40 @@
-use stack_vec::StackVec;
 use console::{kprint, kprintln, CONSOLE};
+use std::str;
+use pi::timer::{current_time, spin_sleep_ms};
+use pi::atags::Atags;
+#[cfg(not(test))]
+use ALLOCATOR;
 
 /// Error type for `Command` parse failures.
 #[derive(Debug)]
 enum Error {
     Empty,
-    TooManyArgs
 }
 
+const BEL: u8 = 0x07u8;
+const BS: u8 = 0x08u8;
+const LF: u8 = 0x0au8;
+const CR: u8 = 0x0du8;
+const ESC: u8 = 0x1bu8;
+const DEL: u8 = 0x7fu8;
+
+const BANNER: &str = r#"
+                 _ _ ___         
+             __-         `-_    
+         ___/__        〇   \ 
+     - '     _/             /
+   '_'             /
+ / _- ---            __ -
+/`     |          _ / \  \
+       |       -       \ |
+        \    /          V
+          \  |
+            \ \
+              \"#;
+
 /// A structure representing a single shell command.
 struct Command<'a> {
-    args: StackVec<'a, &'a str>
+    args: Vec<&'a str>,
 }
 
 impl<'a> Command<'a> {
@@ -21,10 +45,10 @@ impl<'a> Command<'a> {
     ///
     /// If `s` contains no arguments, returns `Error::Empty`. If there are more
     /// arguments than `buf` can hold, returns `Error::TooManyArgs`.
-    fn parse(s: &'a str, buf: &'a mut [&'a str]) -> Result<Command<'a>, Error> {
-        let mut args = StackVec::new(buf);
+    fn parse(s: &str) -> Result<Command, Error> {
+        let mut args = Vec::with_capacity(64);
         for arg in s.split(' ').filter(|a| !a.is_empty()) {
-            args.push(arg).map_err(|_| Error::TooManyArgs)?;
+            args.push(arg);
         }
 
         if args.is_empty() {
@@ -36,12 +60,229 @@ impl<'a> Command<'a> {
 
     /// Returns this command's path. This is equivalent to the first argument.
     fn path(&self) -> &str {
-        unimplemented!()
+        self.args[0]
     }
 }
 
 /// Starts a shell using `prefix` as the prefix for each line. This function
 /// never returns: it is perpetually in a shell loop.
 pub fn shell(prefix: &str) -> ! {
-    unimplemented!()
+    spin_sleep_ms(200); // wait a little time for client to attach
+    kprintln!("{}", BANNER);
+
+    let mut history = Vec::new();
+    loop {
+        kprint!("{}", prefix);
+
+        let line = read_line(&mut history);
+        match Command::parse(&line) {
+            Ok(command) => {
+                let path = command.path();
+                match path {
+                    "atags" => atags(&command),
+                    "echo" => echo(&command),
+                    "echohex" => echohex(&command),
+                    "exit" => exit(&command),
+                    "memstat" => memstat(&command),
+                    "panic" => panic(&command),
+                    "uptime" => uptime(&command),
+                    _ => kprintln!("unknown command: {}", path),
+                }
+            }
+            Err(Error::Empty) => {
+                // Ignore
+            }
+        }
+    }
+}
+
+fn read_line(history: &mut Vec<Vec<u8>>) -> String {
+    let mut console = CONSOLE.lock();
+    let mut cursor = 0;
+    let mut line_vec = Vec::with_capacity(512);
+    let mut history_index = history.len();
+    loop {
+        match console.read_byte() {
+            BS | DEL => {
+                // Backspace
+                if cursor > 0 {
+                    cursor -= 1;
+                    line_vec.remove(cursor);
+
+                    console.write_byte(BS);
+                    for byte in &line_vec[cursor..] {
+                        console.write_byte(*byte);
+                    }
+                    console.write_byte(b' ');
+                    for _i in cursor..line_vec.len() {
+                        console.write_byte(BS);
+                    }
+                    console.write_byte(BS);
+                } else {
+                    console.write_byte(BEL);
+                }
+            }
+            CR | LF => {
+                // Return
+                console.write_byte(CR);
+                console.write_byte(LF);
+                break;
+            }
+            ESC => {
+                match console.read_byte() {
+                    b'[' => {
+                        match console.read_byte() {
+                            b'D' => {
+                                // Left arrow
+                                if cursor > 0 {
+                                    cursor -= 1;
+                                    console.write_byte(ESC);
+                                    console.write_byte(b'[');
+                                    console.write_byte(b'D');
+                                } else {
+                                    console.write_byte(BEL);
+                                }
+                            }
+                            b'C' => {
+                                // Right arrow
+                                if cursor < line_vec.len() {
+                                    cursor += 1;
+                                    console.write_byte(ESC);
+                                    console.write_byte(b'[');
+                                    console.write_byte(b'C');
+                                } else {
+                                    console.write_byte(BEL);
+                                }
+                            }
+                            direction @ b'A' | direction @ b'B' => {
+                                if direction == b'A' && history_index > 0 {
+                                    // Up arrow
+                                    history_index -= 1;
+                                } else if direction == b'B' && history_index < history.len() - 1 {
+                                    // Down arrow
+                                    history_index += 1;
+                                } else {
+                                    console.write_byte(BEL);
+                                    continue;
+                                }
+
+                                cursor = line_vec.len();
+                                for _ in 0..line_vec.len() {
+                                    console.write_byte(BS);
+                                }
+                                for _ in 0..line_vec.len() {
+                                    console.write_byte(b' ');
+                                }
+                                for _ in 0..line_vec.len() {
+                                    console.write_byte(BS);
+                                }
+                                line_vec = history[history_index].clone();
+                                for byte in &line_vec {
+                                    console.write_byte(*byte);
+                                }
+                            }
+                            _ => {
+                                console.write_byte(BEL);
+                            }
+                        }
+                    }
+                    _ => {
+                        console.write_byte(BEL);
+                    }
+                }
+            }
+            byte if byte.is_ascii_graphic() || byte == b' ' => {
+                line_vec.insert(cursor, byte);
+                for byte in &line_vec[cursor..] {
+                    console.write_byte(*byte);
+                }
+                cursor += 1;
+                for _i in cursor..line_vec.len() {
+                    console.write_byte(BS);
+                }
+            }
+            _ => {
+                // unrecognized characters
+                console.write_byte(BEL);
+            }
+        }
+    }
+
+    history.push(line_vec.clone());
+    String::from_utf8(line_vec).unwrap_or_default()
+}
+
+fn echo(command: &Command) {
+    if command.args.len() > 1 {
+        kprint!("{}", command.args[1]);
+        if command.args.len() > 2 {
+            for arg in &command.args[2..] {
+                kprint!(" {}", *arg);
+            }
+        }
+    }
+
+    kprintln!();
+}
+
+fn echohex(command: &Command) {
+    echo(&command);
+    if command.args.len() > 1 {
+        for byte in command.args[1].bytes() {
+            kprint!("{:02X}", byte);
+        }
+        if command.args.len() > 2 {
+            for arg in &command.args[2..] {
+                kprint!(" ");
+                for byte in arg.bytes() {
+                    kprint!("{:02X}", byte);
+                }
+            }
+        }
+    }
+
+    kprintln!();
+}
+
+fn uptime(_command: &Command) {
+    let time = current_time();
+    let sec = time / 1000 / 1000;
+    let min = sec / 60;
+    let hour = min / 60;
+    let day = hour / 24;
+    kprintln!(
+        "Uptime: {} day(s), {:02}:{:02}:{:02}",
+        day,
+        hour % 24,
+        min % 60,
+        sec % 60
+    );
+}
+
+fn exit(_command: &Command) -> ! {
+    kprintln!("Might be dangerous: bootloader might be corrupted.");
+    kprintln!("Now exiting to bootloader. Detach and send a new kernel to continue.");
+    // in line with bootloader
+    const BOOTLOADER_START_ADDR: usize = 0x4000000;
+    unsafe {
+        asm!("br $0" : : "r"(BOOTLOADER_START_ADDR));
+        loop {
+            asm!("nop" :::: "volatile")
+        }
+    }
+}
+
+fn panic(_command: &Command) -> ! {
+    panic!("You ask me to panic!");
+}
+
+fn atags(_command: &Command) {
+    for atag in Atags::get() {
+        kprintln!("{:?}", atag);
+    }
+}
+
+fn memstat(_command: &Command) {
+    #[cfg(not(test))]
+    kprintln!("Allocator: {:?}", ALLOCATOR);
 }
diff --git a/pi/src/atags/atag.rs b/pi/src/atags/atag.rs
index 641d2f2..78025eb 100644
--- a/pi/src/atags/atag.rs
+++ b/pi/src/atags/atag.rs
@@ -1,4 +1,7 @@
 use atags::raw;
+use std::str;
+use std::ffi::CStr;
+use std::os::raw::c_char;
 
 pub use atags::raw::{Core, Mem};
 
@@ -9,24 +12,33 @@ pub enum Atag {
     Mem(raw::Mem),
     Cmd(&'static str),
     Unknown(u32),
-    None
+    None,
 }
 
 impl Atag {
     /// Returns `Some` if this is a `Core` ATAG. Otherwise returns `None`.
     pub fn core(self) -> Option<Core> {
-        unimplemented!()
+        match self {
+            Atag::Core(core) => Some(core),
+            _ => None,
+        }
     }
 
     /// Returns `Some` if this is a `Mem` ATAG. Otherwise returns `None`.
     pub fn mem(self) -> Option<Mem> {
-        unimplemented!()
+        match self {
+            Atag::Mem(mem) => Some(mem),
+            _ => None,
+        }
     }
 
     /// Returns `Some` with the command line string if this is a `Cmd` ATAG.
     /// Otherwise returns `None`.
     pub fn cmd(self) -> Option<&'static str> {
-        unimplemented!()
+        match self {
+            Atag::Cmd(cmd) => Some(cmd),
+            _ => None,
+        }
     }
 }
 
@@ -40,11 +52,14 @@ impl<'a> From<&'a raw::Atag> for Atag {
 
         unsafe {
             match (atag.tag, &atag.kind) {
-                (raw::Atag::CORE, &raw::Kind { core }) => unimplemented!(),
-                (raw::Atag::MEM, &raw::Kind { mem }) => unimplemented!(),
-                (raw::Atag::CMDLINE, &raw::Kind { ref cmd }) => unimplemented!(),
-                (raw::Atag::NONE, _) => unimplemented!(),
-                (id, _) => unimplemented!()
+                (raw::Atag::CORE, &raw::Kind { core }) => Atag::Core(core),
+                (raw::Atag::MEM, &raw::Kind { mem }) => Atag::Mem(mem),
+                (raw::Atag::CMDLINE, &raw::Kind { ref cmd }) => {
+                    let cmdline = CStr::from_ptr((&cmd.cmd as *const u8) as *const c_char);
+                    Atag::Cmd(str::from_utf8_unchecked(cmdline.to_bytes()))
+                }
+                (raw::Atag::NONE, _) => Atag::None,
+                (id, _) => Atag::Unknown(id),
             }
         }
     }
diff --git a/pi/src/atags/mod.rs b/pi/src/atags/mod.rs
index 97b9776..d20bc56 100644
--- a/pi/src/atags/mod.rs
+++ b/pi/src/atags/mod.rs
@@ -24,6 +24,13 @@ impl Iterator for Atags {
     type Item = Atag;
 
     fn next(&mut self) -> Option<Atag> {
-        unimplemented!("atags iterator")
+        // Skip the first CORE ATAG
+        match self.ptr.next() {
+            Some(next_atag) => {
+                self.ptr = next_atag;
+                Some(next_atag.into())
+            }
+            None => None
+        }
     }
 }
diff --git a/pi/src/atags/raw.rs b/pi/src/atags/raw.rs
index 19162c3..891e209 100644
--- a/pi/src/atags/raw.rs
+++ b/pi/src/atags/raw.rs
@@ -3,7 +3,7 @@
 pub struct Atag {
     pub dwords: u32,
     pub tag: u32,
-    pub kind: Kind
+    pub kind: Kind,
 }
 
 impl Atag {
@@ -20,7 +20,15 @@ impl Atag {
 
     /// Returns the ATAG following `self`, if there is one.
     pub fn next(&self) -> Option<&Atag> {
-        unimplemented!()
+        unsafe {
+            let addr = (self as *const Atag) as *const u32;
+            let next_atag = addr.offset(self.dwords as isize) as *mut Atag;
+
+            match (*next_atag).tag {
+                super::raw::Atag::NONE => None,
+                _ => next_atag.as_ref(),
+            }
+        }
     }
 }
 
@@ -29,7 +37,7 @@ impl Atag {
 pub union Kind {
     pub core: Core,
     pub mem: Mem,
-    pub cmd: Cmd
+    pub cmd: Cmd,
 }
 
 /// A `CORE` ATAG.
@@ -38,7 +46,7 @@ pub union Kind {
 pub struct Core {
     pub flags: u32,
     pub page_size: u32,
-    pub root_dev: u32
+    pub root_dev: u32,
 }
 
 /// A `MEM` ATAG.
@@ -46,7 +54,7 @@ pub struct Core {
 #[derive(Debug, Copy, Clone)]
 pub struct Mem {
     pub size: u32,
-    pub start: u32
+    pub start: u32,
 }
 
 /// A `CMDLINE` ATAG.
@@ -54,5 +62,5 @@ pub struct Mem {
 #[derive(Debug, Copy, Clone)]
 pub struct Cmd {
     /// The first byte of the command line string.
-    pub cmd: u8
+    pub cmd: u8,
 }
diff --git a/pi/src/gpio.rs b/pi/src/gpio.rs
index f70bfb5..f914de1 100644
--- a/pi/src/gpio.rs
+++ b/pi/src/gpio.rs
@@ -102,7 +102,8 @@ impl Gpio<Uninitialized> {
     /// Enables the alternative function `function` for `self`. Consumes self
     /// and returns a `Gpio` structure in the `Alt` state.
     pub fn into_alt(self, function: Function) -> Gpio<Alt> {
-        unimplemented!()
+        self.registers.FSEL[(self.pin / 10) as usize].or_mask((function as u32) << (3 * (self.pin % 10)));
+        self.transition()
     }
 
     /// Sets this pin to be an _output_ pin. Consumes self and returns a `Gpio`
@@ -121,12 +122,12 @@ impl Gpio<Uninitialized> {
 impl Gpio<Output> {
     /// Sets (turns on) the pin.
     pub fn set(&mut self) {
-        unimplemented!()
+        self.registers.SET[(self.pin / 32) as usize].write(1 << (self.pin % 32));
     }
 
     /// Clears (turns off) the pin.
     pub fn clear(&mut self) {
-        unimplemented!()
+        self.registers.CLR[(self.pin / 32) as usize].write(1 << (self.pin % 32));
     }
 }
 
@@ -134,6 +135,6 @@ impl Gpio<Input> {
     /// Reads the pin's value. Returns `true` if the level is high and `false`
     /// if the level is low.
     pub fn level(&mut self) -> bool {
-        unimplemented!()
+        self.registers.LEV[(self.pin / 32) as usize].has_mask(1 << (self.pin % 32))
     }
 }
diff --git a/pi/src/lib.rs b/pi/src/lib.rs
index e6c8952..87b0252 100644
--- a/pi/src/lib.rs
+++ b/pi/src/lib.rs
@@ -2,7 +2,6 @@
 #![feature(const_fn)]
 #![feature(asm)]
 #![feature(decl_macro)]
-#![feature(repr_align)]
 #![feature(attr_literals)]
 #![feature(never_type)]
 #![feature(pointer_methods)]
diff --git a/pi/src/timer.rs b/pi/src/timer.rs
index e0aa184..db60af2 100644
--- a/pi/src/timer.rs
+++ b/pi/src/timer.rs
@@ -30,21 +30,29 @@ impl Timer {
     /// Reads the system timer's counter and returns the 64-bit counter value.
     /// The returned value is the number of elapsed microseconds.
     pub fn read(&self) -> u64 {
-        unimplemented!()
+        let low = self.registers.CLO.read();
+        let high = self.registers.CHI.read();
+        ((high as u64) << 32) | (low as u64)
     }
 }
 
 /// Returns the current time in microseconds.
 pub fn current_time() -> u64 {
-    unimplemented!()
+    Timer::new().read()
 }
 
 /// Spins until `us` microseconds have passed.
 pub fn spin_sleep_us(us: u64) {
-    unimplemented!()
+    let old = current_time();
+    loop {
+        let new = current_time();
+        if old + us <= new {
+            break;
+        }
+    }
 }
 
 /// Spins until `ms` milliseconds have passed.
 pub fn spin_sleep_ms(ms: u64) {
-    unimplemented!()
+    spin_sleep_us(ms * 1000);
 }
diff --git a/pi/src/uart.rs b/pi/src/uart.rs
index 92b6159..4d00b59 100644
--- a/pi/src/uart.rs
+++ b/pi/src/uart.rs
@@ -24,6 +24,26 @@ enum LsrStatus {
 #[allow(non_snake_case)]
 struct Registers {
     // FIXME: Declare the "MU" registers from page 8.
+    AUX_MU_IO_REG: Volatile<u8>,
+    __r0: [Reserved<u8>; 3],
+    AUX_MU_IER_REG: Volatile<u8>,
+    __r1: [Reserved<u8>; 3],
+    AUX_MU_IIR_REG: Volatile<u8>,
+    __r2: [Reserved<u8>; 3],
+    AUX_MU_LCR_REG: Volatile<u8>,
+    __r3: [Reserved<u8>; 3],
+    AUX_MU_MCR_REG: Volatile<u8>,
+    __r4: [Reserved<u8>; 3],
+    AUX_MU_LSR_REG: ReadVolatile<u8>,
+    __r5: [Reserved<u8>; 3],
+    AUX_MU_MSR_REG: ReadVolatile<u8>,
+    __r6: [Reserved<u8>; 3],
+    AUX_MU_SCRATCH: Volatile<u8>,
+    __r7: [Reserved<u8>; 3],
+    AUX_MU_CNTL_REG: Volatile<u8>,
+    __r8: [Reserved<u8>; 3],
+    AUX_MU_STAT_REG: ReadVolatile<u32>,
+    AUX_MU_BAUD: Volatile<u16>
 }
 
 /// The Raspberry Pi's "mini UART".
@@ -46,27 +66,41 @@ impl MiniUart {
             (*AUX_ENABLES).or_mask(1);
             &mut *(MU_REG_BASE as *mut Registers)
         };
-
+        
         // FIXME: Implement remaining mini UART initialization.
-        unimplemented!()
+        // GPIO 14 15 set to TX/RX
+        Gpio::new(14).into_alt(Function::Alt5);
+        Gpio::new(15).into_alt(Function::Alt5);
+        registers.AUX_MU_LCR_REG.or_mask(0b11); // 8-bit data size
+        // baudrate = system_clock_freq / (8 * (baud_reg + 1))
+        // system_clock_freq is 250MHz (Page 10)
+        // 250MHz / 8 / 115200 - 1 = 270
+        registers.AUX_MU_BAUD.write(270); // 16-bit baudrate register
+        registers.AUX_MU_CNTL_REG.or_mask(0b11); // enable TX & RX
+
+        MiniUart {
+            registers: registers,
+            timeout: None,
+        }
     }
 
     /// Set the read timeout to `milliseconds` milliseconds.
     pub fn set_read_timeout(&mut self, milliseconds: u32) {
-        unimplemented!()
+        self.timeout = Some(milliseconds)
     }
 
     /// Write the byte `byte`. This method blocks until there is space available
     /// in the output FIFO.
     pub fn write_byte(&mut self, byte: u8) {
-        unimplemented!()
+        while !self.registers.AUX_MU_LSR_REG.has_mask(LsrStatus::TxAvailable as u8) { };
+        self.registers.AUX_MU_IO_REG.write(byte);
     }
 
     /// Returns `true` if there is at least one byte ready to be read. If this
     /// method returns `true`, a subsequent call to `read_byte` is guaranteed to
     /// return immediately. This method does not block.
     pub fn has_byte(&self) -> bool {
-        unimplemented!()
+        self.registers.AUX_MU_LSR_REG.has_mask(LsrStatus::DataReady as u8)
     }
 
     /// Blocks until there is a byte ready to read. If a read timeout is set,
@@ -78,17 +112,50 @@ impl MiniUart {
     /// returns `Ok(())`, a subsequent call to `read_byte` is guaranteed to
     /// return immediately.
     pub fn wait_for_byte(&self) -> Result<(), ()> {
-        unimplemented!()
+        match self.timeout {
+            None => {
+                loop {
+                    if self.has_byte() {
+                        return Ok(())
+                    }
+                }
+            },
+            Some(timeout) => {
+                let old = timer::current_time();
+                let deadline = old + (timeout as u64) * 1000; // Internal clock unit is microsecond, but timeout unit is millisecond
+                while deadline >= timer::current_time() {
+                    if self.has_byte() {
+                        return Ok(())
+                    }
+                }
+                Err(())
+            }
+        }
     }
 
     /// Reads a byte. Blocks indefinitely until a byte is ready to be read.
     pub fn read_byte(&mut self) -> u8 {
-        unimplemented!()
+        while !self.has_byte() { }
+        self.registers.AUX_MU_IO_REG.read()
     }
 }
 
 // FIXME: Implement `fmt::Write` for `MiniUart`. A b'\r' byte should be written
 // before writing any b'\n' byte.
+impl fmt::Write for MiniUart {
+    fn write_str(&mut self, s: &str) -> fmt::Result {
+        for byte in s.bytes() {
+            match byte {
+                b'\n' => {
+                    self.write_byte(b'\r');
+                    self.write_byte(b'\n');
+                },
+                _ => self.write_byte(byte)
+            }
+        }
+        Ok(())
+    }
+}
 
 #[cfg(feature = "std")]
 mod uart_io {
@@ -104,4 +171,34 @@ mod uart_io {
     //
     // The `io::Write::write()` method must write all of the requested bytes
     // before returning.
+    impl io::Read for MiniUart {
+       fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+           match self.wait_for_byte() {
+               Ok(()) => {
+                   let mut index = 0;
+                   while self.has_byte() && index < buf.len() {
+                       buf[index] = self.read_byte();
+                       index += 1;
+                   }
+                   Ok(index)
+               },
+               Err(()) => {
+                   Err(io::Error::new(io::ErrorKind::TimedOut, "reading UART timed out"))
+               }
+           }
+       }
+    }
+
+    impl io::Write for MiniUart {
+       fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+           for byte in buf {
+               self.write_byte(*byte);
+           }
+           Ok(buf.len())
+       }
+
+       fn flush(&mut self) -> io::Result<()> {
+           unimplemented!()
+       }
+    }
 }
diff --git a/std/src/collections/hash/map.rs b/std/src/collections/hash/map.rs
index 4114354..ca6c7c9 100644
--- a/std/src/collections/hash/map.rs
+++ b/std/src/collections/hash/map.rs
@@ -1972,7 +1972,7 @@ impl<'a, K, V> Placer<V> for Entry<'a, K, V> {
 #[unstable(feature = "collection_placement",
            reason = "placement protocol is subject to change",
            issue = "30172")]
-impl<'a, K, V> Place<V> for EntryPlace<'a, K, V> {
+unsafe impl<'a, K, V> Place<V> for EntryPlace<'a, K, V> {
     fn pointer(&mut self) -> *mut V {
         self.bucket.read_mut().1
     }
